File: core/py/src/archon_core/hexbytes.py
"""Typed, fail-closed hex spellings.

Not a hex helper — every language has one — but the FIXED SIZE. Each decoder accepts exactly
its type's byte count or fails; the length is checked on the TEXT before any byte is decoded,
so a wrong-length input never reaches the decoder. Either case in, lowercase out. No `0x`, no
whitespace, no odd-length bodies.
"""

from .crypto import PUBLIC_KEY_SIZE, SEED_SIZE, SIGNATURE_SIZE

__all__ = ["to_hex", "seed_from_hex", "pubkey_from_hex", "signature_from_hex"]

_HEX_DIGITS = frozenset("0123456789abcdefABCDEF")


def to_hex(data: bytes) -> str:
    """Lowercase hex, no prefix."""
    return bytes(data).hex()


def _fixed(text: str, n: int, what: str) -> bytes:
    if len(text) != n * 2:
        raise ValueError(
            f"hexbytes: {what} hex is {len(text)} characters, expected {n * 2}"
        )
    # Python's bytes.fromhex accepts whitespace; archon's grammar does not. Check the
    # alphabet explicitly rather than relying on the decoder's laxer rules — the other
    # cores reject a space here, and agreement is the point.
    for ch in text:
        if ch not in _HEX_DIGITS:
            raise ValueError(f"hexbytes: {what}: non-hex character {ch!r}")
    return bytes.fromhex(text)


def seed_from_hex(text: str) -> bytes:
    """Exactly 32 bytes, or raises."""
    return _fixed(text, SEED_SIZE, "seed")


def pubkey_from_hex(text: str) -> bytes:
    """Exactly 32 bytes, or raises."""
    return _fixed(text, PUBLIC_KEY_SIZE, "public key")


def signature_from_hex(text: str) -> bytes:
    """Exactly 64 bytes, or raises."""
    return _fixed(text, SIGNATURE_SIZE, "signature")

File: core/java/src/main/java/dev/bitspark/archon/core/HexBytes.java
package dev.bitspark.archon.core;

/**
 * Typed, fail-closed hex spellings.
 *
 * <p>Not a hex helper — every language has one — but the FIXED SIZE. Each decoder accepts exactly
 * its type's byte count or fails, and the length is checked on the TEXT before any byte is
 * decoded. Either case in, lowercase out. No {@code 0x}, no whitespace, no odd-length bodies.
 */
public final class HexBytes {

  private HexBytes() {}

  /** Lowercase hex, no prefix. */
  public static String toHex(byte[] data) {
    StringBuilder out = new StringBuilder(data.length * 2);
    for (byte b : data) {
      out.append(Character.forDigit((b >> 4) & 0xF, 16));
      out.append(Character.forDigit(b & 0xF, 16));
    }
    return out.toString();
  }

  /** Exactly 32 bytes, or throws. */
  public static byte[] seedFromHex(String text) {
    return fixed(text, Crypto.SEED_SIZE, "seed");
  }

  /** Exactly 32 bytes, or throws. */
  public static byte[] pubkeyFromHex(String text) {
    return fixed(text, Crypto.PUBLIC_KEY_SIZE, "public key");
  }

  /** Exactly 64 bytes, or throws. */
  public static byte[] signatureFromHex(String text) {
    return fixed(text, Crypto.SIGNATURE_SIZE, "signature");
  }

  private static byte[] fixed(String text, int n, String what) {
    if (text.length() != n * 2) {
      throw new IllegalArgumentException(
          "hexbytes: " + what + " hex is " + text.length() + " characters, expected " + n * 2);
    }
    byte[] out = new byte[n];
    for (int i = 0; i < n; i++) {
      int hi = digit(text.charAt(2 * i), what);
      int lo = digit(text.charAt(2 * i + 1), what);
      out[i] = (byte) ((hi << 4) | lo);
    }
    return out;
  }

  /**
   * One hex digit, or a refusal.
   *
   * <p>{@code Character.digit} would also accept non-ASCII digits and other radix-16 characters
   * the grammar does not allow, so the alphabet is checked explicitly. The other cores reject
   * those, and agreement is the point.
   */
  private static int digit(char c, String what) {
    if (c >= '0' && c <= '9') {
      return c - '0';
    }
    if (c >= 'a' && c <= 'f') {
      return c - 'a' + 10;
    }
    if (c >= 'A' && c <= 'F') {
      return c - 'A' + 10;
    }
    throw new IllegalArgumentException("hexbytes: " + what + ": non-hex character '" + c + "'");
  }
}
